core/server.go

```diff
@@ -5,21 +5,102 @@ package core
 import (
 	"context"
 
+	"github.com/anz-bank/sysl-go/config"
 	"github.com/prometheus/client_golang/prometheus"
 	"github.com/sirupsen/logrus"
 )
 
+type ServerParams struct {
+	Ctx                context.Context
+	Name               string
+	Config             *config.DefaultConfig
+	logger             *logrus.Logger
+	restManager        RestManager
+	grpcManager        GrpcManager
+	prometheusRegistry *prometheus.Registry
+}
+
+func NewServerParams(ctx context.Context, name string, config *config.DefaultConfig) *ServerParams {
+	return &ServerParams{Ctx: ctx, Name: name, Config: config}
+}
+
+//nolint:gocognit // Long method names are okay because only generated code will call this, not humans.
+func (params *ServerParams) Start(opts ...ServerOption) error {
+	for _, o := range opts {
+		if err := o.Apply(params); err != nil {
+			return err
+		}
+	}
+	return Server(params)
+}
+
+type ServerOption interface {
+	Apply(params *ServerParams) error
+}
+
+type restManagerOption struct {
+	restManager RestManager
+}
+
+func (o *restManagerOption) Apply(params *ServerParams) error {
+	params.restManager = o.restManager
+	return nil
+}
+
+func WithRestManager(manager RestManager) ServerOption {
+	return &restManagerOption{manager}
+}
+
+type loggerOption struct {
+	logger *logrus.Logger
+}
+
+func (o *loggerOption) Apply(params *ServerParams) error {
+	params.logger = o.logger
+	return nil
+}
+
+func WithLogger(logger *logrus.Logger) ServerOption {
+	return &loggerOption{logger}
+}
+
+func WithPrometheusRegistry(prometheusRegistry *prometheus.Registry) ServerOption {
+	return &prometheusRegistryOption{prometheusRegistry}
+}
+
+type prometheusRegistryOption struct {
+	prometheusRegistry *prometheus.Registry
+}
+
+func (o *prometheusRegistryOption) Apply(params *ServerParams) error {
+	params.prometheusRegistry = o.prometheusRegistry
+	return nil
+}
+
+type grpcManagerOption struct {
+	grpcManager GrpcManager
+}
+
+func (o *grpcManagerOption) Apply(params *ServerParams) error {
+	params.grpcManager = o.grpcManager
+	return nil
+}
+
+func WithGrpcManager(manager GrpcManager) ServerOption {
+	return &grpcManagerOption{manager}
+}
+
 //nolint:gocognit // Long method names are okay because only generated code will call this, not humans.
-func Server(ctx context.Context, name string, hl Manager, grpcHl GrpcManager, logger *logrus.Logger, promRegistry *prometheus.Registry) error {
-	mWare := prepareMiddleware(name, logger, promRegistry)
+func Server(params *ServerParams) error {
```

Comment:
I'd like this function signature to stay as it was to keep backwards compatibility. Internally it should route functionality to `ServerParams`:

```go
func Server(ctx context.Context, name string, hl Manager, grpcHl GrpcManager, logger *logrus.Logger, promRegistry *prometheus.Registry) error {
    return NewServerParams(ctx, name,
            WithLogrusLogger(logrus),
            WithRestManager(hl),
            WithGrpcManager(grpcHl),
            WithPrometheusRegistry(promRegistry))
        .Start()
```